cloud-storage-uploader/src/client.ts

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as fs from 'fs';
+import { UploadHelper } from './upload-helper';
+import { Storage, UploadResponse } from '@google-cloud/storage';
+
+/**
+ * Available options to create the client.
+ *
+ * @param credentials GCP JSON credentials (default uses ADC).
+ */
+
+type ClientOptions = {
+  credentials?: string;
+};
+
+/**
+ * Handles credential lookup, registration and
+ * wraps interactions with the GCS Helper
+ *
+ * @param opts list of ClientOptions
+ * @returns Client
+ */
+export class Client {
+  readonly storage: Storage;
+  constructor(opts?: ClientOptions) {
+    if (opts?.credentials) {
+      // If the credentials are not JSON, they are probably base64-encoded. Even
+      // though we don't instruct users to provide base64-encoded credentials,
+      // sometimes they still do.
+      if (!opts.credentials.trim().startsWith('{')) {
+        const creds = opts.credentials;
+        opts.credentials = Buffer.from(creds, 'base64').toString('utf8');
+      }
+      const creds = JSON.parse(opts.credentials);
+      this.storage = new Storage({ credentials: creds });
+    } else {
+      this.storage = new Storage();
+    }
+  }
+  /**
+   * Invokes GCS Helper for uploading file
+   * or directory
+   * @param bucketName Name of bucket to upload file/dir
+   * @param path path of the file/dir to upload
+   * @param prefix optional prefix when uploading to GCS
+   */
+  async upload(destination: string, path: string): Promise<UploadResponse[]> {
+    let bucketName = destination;
+    let prefix = '';
+    //if destination of the form my-bucket/subfolder
+    // get bucket and prefix
+    if (destination.indexOf('/') > -1) {
```

Comment:
It'd be slightly more performant if you did:

```node
idx := destination.indexOf('/')
if (idx > -1) {
  bucketName = destination.substring(0, idx)
  prefix = destination.substring(idx+1);
}
```

Reply:
Hmm, I first thought this meant the walrus operator `:=` was available in TS, but that doesn't seem to be the case.
I have made idx to a const and followed this, but lmk if you meant the the walrus operator